Add Ecotone Fee Information to JSON Receipts

trianglesphere opened the pull request:
**Description**

This commit adds the following fields the receipt:
- L1 Blob BaseFee
- L1 BaseFee Scalar
- L1 BlobBaseFee Scalar

The new fields are only present after Ecotone. The field `feeScalar` which is pre-ecotone is set to null now for receipts that occurred after Ecotone's activation.

I also refactored extractL1GasParams to return a struct rather than a list of value due to the large number of returned fields.

**Tests**

I modified the post-ecotone unit tests to check for these fields.

**Metadata**

- Fixes https://github.com/ethereum-optimism/client-pod/issues/640

sebastianst commented on core/types/receipt.go:
```diff
@@ -84,11 +84,14 @@ type Receipt struct {
-	L1GasUsed  *big.Int   `json:"l1GasUsed,omitempty"`
-	L1Fee      *big.Int   `json:"l1Fee,omitempty"`
-	FeeScalar  *big.Float `json:"l1FeeScalar,omitempty"` // always nil after Ecotone hardfork
+	// OVM legacy & Bedrock: extend receipts with their L1 price (if a rollup tx)
```
```suggestion
	// Optimism: extend receipts with L1 fee info
```

sebastianst commented on core/types/rollup_cost.go:
```diff
@@ -218,37 +218,63 @@ func newL1CostFuncEcotone(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseF
+		costFunc:            costFunc,
+		feeScalar:           feeScalar,
+		l1BaseFeeScalar:     nil,
+		l1BlobBaseFeeScalar: nil,
```
could skip the `nil`s

sebastianst commented on core/types/receipt_test.go:
```diff
@@ -698,9 +698,78 @@ func clearComputedFieldsOnLogs(logs []*Log) []*Log {
-	t *testing.T, l1AttributesPayload []byte,
-	l1GasPrice, l1GasUsed *big.Int, feeScalar *big.Float, l1Fee *big.Int) ([]*Transaction, []*Receipt) {
+func getOptimismEcotoneTxReceipts(l1AttributesPayload []byte, l1GasPrice, l1BlobGasPrice, l1GasUsed, l1Fee *big.Int, baseFeeScalar, blobBaseFeeScalar *big.Float) ([]*Transaction, []*Receipt) {
+	//to4 := common.HexToAddress("0x4")
```
delete

sebastianst commented on core/types/receipt.go:
```diff
@@ -84,11 +84,14 @@ type Receipt struct {
+	L1GasUsed           *big.Int   `json:"l1GasUsed,omitempty"`
+	L1Fee               *big.Int   `json:"l1Fee,omitempty"`
+	FeeScalar           *big.Float `json:"l1FeeScalar,omitempty"`         // always nil after Ecotone hardfork
+	L1BaseFeeScalar     *big.Float `json:"l1BaseFeeScalar,omitempty"`     // always nil prior to the Ecotone hardfork
+	L1BlobBaseFeeScalar *big.Float `json:"l1BlobBaseFeeScalar,omitempty"` // always nil prior to the Ecotone hardfork
```
Why don't we make them `*big.Int`s or even `uint32`s? By the [Ecotone spec](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#ecotone-l1-cost-fee-changes-eip-4844-da) they fit into `uint32`s. This would allow us to avoid dealing with floats and use exact values instead.

trianglesphere replied:
For the fee scalar this was intentional to keep them as floats, but we can switch to `*f64` from `*big.Float` if we want. I talked with @MSilb7 & pre-ecotone, we already have floats & it is easier to use rather than having to recompute by dividing by 1e6.

MSilb7 replied:
Yeah the original intent was just to stay consistent with the legacy scalar format, since there was no strong opinion either way.

Since these are net new (so no breaking changes), either format works, as long as it's clear. So if we now have a strong opinion about not being float, that would work.

sebastianst commented on core/types/rollup_cost.go:
```diff
@@ -218,37 +218,63 @@ func newL1CostFuncEcotone(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseF
+	l1BlobBaseFee       *big.Int
+	costFunc            l1CostFunc
+	feeScalar           *big.Float // pre-ecotone
+	l1BaseFeeScalar     *big.Float // post-ecotone
+	l1BlobBaseFeeScalar *big.Float // post-ecotone
```
see other comment, I suggest to make them `uint32`s.

sebastianst commented on core/types/rollup_cost.go:
```diff
@@ -218,37 +218,63 @@ func newL1CostFuncEcotone(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseF
 // extractL1GasParams extracts the gas parameters necessary to compute gas costs from L1 block info
-func extractL1GasParams(config *params.ChainConfig, time uint64, data []byte) (l1BaseFee *big.Int, costFunc l1CostFunc, feeScalar *big.Float, err error) {
+func extractL1GasParams(config *params.ChainConfig, time uint64, data []byte) (gasParams, error) {
 	if config.IsEcotone(time) {
 		// edge case: for the very first Ecotone block we still need to use the Bedrock
 		// function. We detect this edge case by seeing if the function selector is the old one
+		// If so, fall through to the pre-ecotone format
 		if len(data) >= 4 && !bytes.Equal(data[0:4], BedrockL1AttributesSelector) {
```
we could `&&` the `if`s

trianglesphere committed:
Add Ecotone Fee Information to JSON Receipts

This commit adds the following fields the receipt:
- L1 Blob BaseFee
- L1 BaseFee Scalar
- L1 BlobBaseFee Scalar

The new fields are only present after Ecotone. The field `feeScalar`
which is pre-ecotone is set to null now for receipts that occurred
after Ecotone's activation.

I also refactored extractL1GasParams to return a struct rather than
a list of value due to the large number of returned fields.

trianglesphere committed:
Switch Ectone Fee Scalars to be *uint32

sebastianst commented on core/types/receipt_test.go:
```diff
@@ -800,11 +867,10 @@ func TestDeriveOptimismBedrockTxReceipts(t *testing.T) {
-	l1GasUsed := ecotoneGas
-	l1Fee := ecotoneFee
-	txs, receipts := getOptimismTxReceipts(t, payload, l1GasPrice, l1GasUsed, nil /*feeScalar*/, l1Fee)
+	// the parameters we use below are defined in rollup_test.go. Scale the fee scalars by 1e6 to account for the float format.
```
```suggestion
	// the parameters we use below are defined in rollup_test.go
```
we removed the scaling again

sebastianst committed:
Revert comment change in core/types/receipt_test.go

sebastianst committed:
Merge branch 'optimism' into jg/ecotone_json_receipt_data
